util/resolver/authorizer.go

```diff
@@ -272,9 +272,14 @@ func (ah *authHandler) doBearerAuth(ctx context.Context) (token string, err erro
 	if r, exist := ah.scopedTokens[scoped]; exist {
 		ah.Unlock()
 		r.Wait()
-		if r.expires.IsZero() || r.expires.After(time.Now()) {
+		if errors.Is(ctx.Err(), context.Canceled) {
+			return "", ctx.Err()
+		}
+		if !errors.Is(r.err, context.Canceled) &&
+			(r.expires.IsZero() || r.expires.After(time.Now())) {
 			return r.token, r.err
 		}
+		// r.err is canceled or token expired. Will authorize again
```

Comment:
I think there is a race in here (also in old code but less likely to hit) when multiple `Wait` return and both proceed to `r := new(authResult)` . The fix should be to put this `if` condition into a for cycle and remove `r` from `ah.scopedTokens` after getting the lock (after checking that it is still valid).

Reply:
Good point - I also added a check to make sure it's the same `r` being removed